qupath-extension-openslide/src/main/java/qupath/lib/images/servers/openslide/OpenSlide.java

```diff
@@ -0,0 +1,491 @@
+        int levelW = (int) Math.round(relativeDS * w);
+        int levelH = (int) Math.round(relativeDS * h);
+
+        // clip to edge of image
+        levelW = (int) Math.min(levelW, getLevelWidth(level) - levelX);
+        levelH = (int) Math.min(levelH, getLevelHeight(level) - levelY);
+        w = (int) Math.round(levelW / relativeDS);
+        h = (int) Math.round(levelH / relativeDS);
+
+        if (debug) {
+            System.out.println("levelW " + levelW + ", levelH " + levelH
+                    + ", baseX " + baseX + ", baseY " + baseY);
+        }
+
+        if (levelW <= 0 || levelH <= 0) {
+            // nothing to draw
+            return;
+        }
+
+        BufferedImage img = new BufferedImage(levelW, levelH,
+                BufferedImage.TYPE_INT_ARGB_PRE);
+
+        int[] data = ((DataBufferInt) img.getRaster().getDataBuffer())
+                .getData();
+
+        paintRegionARGB(data, baseX, baseY, level, img.getWidth(), img
+                .getHeight());
+
+        // g.scale(1.0 / relativeDS, 1.0 / relativeDS);
+        g.drawImage(img, dx, dy, w, h, null);
+
+        if (debug) {
+            System.out.println(img);
+
+            if (debugThingy == 0) {
+                g.setColor(new Color(1.0f, 0.0f, 0.0f, 0.4f));
+                debugThingy = 1;
+            } else {
+                g.setColor(new Color(0.0f, 1.0f, 0.0f, 0.4f));
+                debugThingy = 0;
+            }
+            g.fillRect(dx, dy, w, h);
+        }
+    }
+
+    final boolean debug = false;
+
+    private int debugThingy = 0;
+
+    public BufferedImage createThumbnailImage(int x, int y, long w, long h,
+                                              int maxSize, int bufferedImageType) throws IOException {
+        double ds;
+
+        if (w > h) {
+            ds = (double) w / maxSize;
+        } else {
+            ds = (double) h / maxSize;
+        }
+
+        if (ds < 1.0) {
+            ds = 1.0;
+        }
+
+        int sw = (int) (w / ds);
+        int sh = (int) (h / ds);
+        int sx = (int) (x / ds);
+        int sy = (int) (y / ds);
+
+        BufferedImage result = new BufferedImage(sw, sh, bufferedImageType);
+
+        Graphics2D g = result.createGraphics();
+        paintRegion(g, 0, 0, sx, sy, sw, sh, ds);
+        g.dispose();
+        return result;
+    }
+
+    public BufferedImage createThumbnailImage(int x, int y, long w, long h,
+                                              int maxSize) throws IOException {
+        return createThumbnailImage(x, y, w, h, maxSize,
+                BufferedImage.TYPE_INT_RGB);
+    }
+
+    public BufferedImage createThumbnailImage(int maxSize) throws IOException {
+        return createThumbnailImage(0, 0, getLevel0Width(), getLevel0Height(),
+                maxSize);
+    }
+
+    public double getLevelDownsample(int level) {
+        return levelDownsamples[level];
+    }
+
+    public int getBestLevelForDownsample(double downsample) {
```

Comment:
The logic here can be problematic sometimes because of how the levels are calculated (see above).

Also, you can end up having a near perfect level match, but still requesting a higher resolution (and many more tiles).

QuPath uses different logic with a bit more tolerance: https://github.com/qupath/qupath/blob/533ce0cf50ca2a5627b55a8c1fe351cdc72e0694/qupath-core/src/main/java/qupath/lib/images/servers/ServerTools.java#L93

In the end, I don't think the method is needed in a minimal wrapper.